kv/dist_sender.go

```diff
@@ -457,23 +459,74 @@ func (ds *DistSender) sendAttempt(trace *tracer.Trace, ba roachpb.BatchRequest,
 func (ds *DistSender) Send(ctx context.Context, ba roachpb.BatchRequest) (*roachpb.BatchResponse, *roachpb.Error) {
 	// In the event that timestamp isn't set and read consistency isn't
 	// required, set the timestamp using the local clock.
-	// TODO(tschottdorf): right place for this?
 	if ba.ReadConsistency == roachpb.INCONSISTENT && ba.Timestamp.Equal(roachpb.ZeroTimestamp) {
-		// Make sure that after the call, args hasn't changed.
-		defer func(timestamp roachpb.Timestamp) {
-			ba.Timestamp = timestamp
-		}(ba.Timestamp)
 		ba.Timestamp = ds.clock.Now()
 	}
 
-	// TODO(tschottdorf): provisional instantiation.
-	return newChunkingSender(ds.sendChunk).Send(ctx, ba)
+	if len(ba.Requests) < 1 {
+		panic("empty batch")
+	}
+
+	splitAndSend := func(ba roachpb.BatchRequest, splitET bool) (*roachpb.BatchResponse, *roachpb.Error, bool) {
+		// Deterministically create ClientCmdIDs for all parts of the batch if
+		// a CmdID is already set (otherwise, leave them empty).
+		var nextID func() roachpb.ClientCmdID
+		empty := roachpb.ClientCmdID{}
+		if empty == ba.CmdID {
+			nextID = func() roachpb.ClientCmdID {
+				return empty
+			}
+		} else {
+			rng := rand.New(rand.NewSource(ba.CmdID.Random))
+			id := ba.CmdID
+			nextID = func() roachpb.ClientCmdID {
+				curID := id             // copy
+				id.Random = rng.Int63() // adjust for next call
+				return curID
+			}
+		}
+
+		var rplChunks []*roachpb.BatchResponse
+		parts := ba.Split(splitET)
+		for _, part := range parts {
+			ba.Requests = part
+			ba.CmdID = nextID()
+			rpl, pErr, shouldSplitET := ds.sendChunk(ctx, ba)
+			if pErr != nil {
+				return nil, pErr, shouldSplitET
+			}
+			// Propagate transaction from last reply to next request. The final
+			// update is taken and put into the response's main header.
+			ba.Txn.Update(rpl.Header().Txn)
+			rplChunks = append(rplChunks, rpl)
+		}
+
+		reply := rplChunks[0]
+		for _, rpl := range rplChunks[1:] {
+			reply.Responses = append(reply.Responses, rpl.Responses...)
+		}
+		lastHeader := rplChunks[len(rplChunks)-1].BatchResponse_Header
+		reply.Error = lastHeader.Error
+		reply.Timestamp = lastHeader.Timestamp
+		reply.Txn = ba.Txn
+		return reply, nil, false
+	}
+
+	rpl, pErr, shouldSplitET := splitAndSend(ba, false /* !splitET */)
```

Comment:
I don't think this works very well.  If `ba.Split()` splits the batch into multiple batches based on commands, then the `shouldSplitET` situation will not be identified until the very last batch is returned, at which point all earlier batches have already been run.  This causes the earlier batches to all be run again, even though they succeeded the first time.

Way up in `client/txn.go`, above the KV package, we enforce that the only place an EndTransaction can appear is the last command in the batch.  I'm assuming this assumption holds in the lower level.

So, all the way back in the `client` package we enforce that an EndTransactionRequest can only appear as the very last command in a batch.  I think this assumption holds in the `kv` layer, although it doesn't appear to be enforced anywhere.

Therefore, my suggestion is to give `sendChunk()` the responsibility of splitting off the EndTransaction() into its own chunk, rather than `batch.Split()`.  If sendChunk() gets a range-spanning chunk, the last command might be an EndTransaction; if it is, run that command in its own batch.